File: lib/features/coupon/domain/models/coupon_model.dart
class CouponModel {
  int? id;
  String? title;
  String? code;
  String? startDate;
  String? expireDate;
  double? minPurchase;
  double? maxDiscount;
  double? discount;
  String? discountType;
  int? status;
  String? createdAt;
  String? updatedAt;
  String? couponType;

  CouponModel(
      {this.id,
        this.title,
        this.code,
        this.startDate,
        this.expireDate,
        this.minPurchase,
        this.maxDiscount,
        this.discount,
        this.discountType,
        this.status,
        this.createdAt,
        this.updatedAt,
        this.couponType,
      });

  CouponModel.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    title = json['title'];
    code = json['code'];
    startDate = json['start_date'];
    expireDate = json['expire_date'];
    minPurchase = json['min_purchase'].toDouble();
    maxDiscount = json['max_discount'].toDouble();
    discount = json['discount'].toDouble();
    discountType = json['discount_type'];
    status = json['status'];
    createdAt = json['created_at'];
    updatedAt = json['updated_at'];
    couponType = json['coupon_type'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['id'] = id;
    data['title'] = title;
    data['code'] = code;
    data['start_date'] = startDate;
    data['expire_date'] = expireDate;
    data['min_purchase'] = minPurchase;
    data['max_discount'] = maxDiscount;
    data['discount'] = discount;
    data['discount_type'] = discountType;
    data['status'] = status;
    data['created_at'] = createdAt;
    data['updated_at'] = updatedAt;
    data['coupon_type'] = couponType;
    return data;
  }
}

class CouponApplyModel {
  final double  discount;
  final String  discountType;

  CouponApplyModel(this.discount, this.discountType);
}

File: models/CouponModel.js
// Assuming you're using a database or ORM like Mongoose
const mongoose = require('mongoose');

// Define the schema for the Coupon model
const couponSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true
  },
  code: {
    type: String,
    required: true
  },
  start_date: {
    type: Date,
    required: true
  },
  expire_date: {
    type: Date,
    required: true
  },
  min_purchase: {
    type: Number,
    required: true
  },
  max_discount: {
    type: Number,
    default: 0
  },
  discount: {
    type: Number,
    required: true
  },
  discount_type: {
    type: String,
    enum: ['amount', 'percent'], // Assuming discount type can be either 'amount' or 'percentage'
    required: true
  },
  status: {
    type: Number,
    default: 1
  },
  created_at: {
    type: Date,
    default: Date.now
  },
  updated_at: {
    type: Date,
    default: Date.now
  },
  coupon_type: {
    type: String,
    enum: ['default'], // Assuming coupon type can be 'default' only
    default: 'default'
  },
  limit: {
    type: Number,
    default: 1
  },
  customer_id: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Customer' // Assuming there's a Customer model related to this coupon
  }
});

// Create the Coupon model from the schema
const Coupon = mongoose.model('Coupon', couponSchema);

// Export the Coupon model
module.exports = Coupon;
